### DISCUSSION TITLE
How to check if a SMILES has a fused ring

### DISCUSSION THREAD

**Discussion by syedzayyan** [rdkit/rdkit#6762] [FAQ] [answered]
Hello All,

I am not sure this could be done with SMART query. I just want to check if a compound has a fused ring and the number of rings in that fused ring. Is that possible with RDKit?

**Comment by bp-kelley** [comment, marked as answer]
Smarts patterns can handle this, you want atoms in rings and not in just one ring:

```
m = Chem.MolFromSmiles("C1CC2CCC21")
pat = Chem.MolFromSmarts("[R&!R1]")
m.GetSubstructMatches(pat)
```

![image](https://github.com/rdkit/rdkit/assets/9888416/3aec0820-3afc-44dd-aaee-a0c8955d585a)

**Reply by syedzayyan** [reply]
you're a genius

Edit:
Thank You!

**Comment by DavidACosgrove** [comment]
I think it might be safer to use the SMARTS ‘x’, number of ring bonds incident on an atom. That way you don’t have to worry about SSSR issues at all. In this case I think it would be [x3,x4] for an atom at the fusion point. My own prejudice is to avoid R if at all possible.

**Reply by syedzayyan** [reply]
why not use R? I have this database of molecules and I want to see how many have a fused ring and how many rings are in this fused ring and then populate the database with that.

**Reply by DavidACosgrove** [reply]
Rn specifies that an atom is in n rings in the Smallest Set of Smallest Rings (SSSR).  The SSSR is not uniquely defined for a molecule, and may not include all the rings, especially in larger fused systems.  Cuban is the classic example.  This article https://depth-first.com/articles/2020/08/31/a-smallest-set-of-smallest-rings/ gives a very reasoned account of the problems.  It links to this article https://www.ics.uci.edu/~dock/manuals/oechem/pyprog/node123.html which gives a more succinct and stronger view.  The upshot is that R isn't a good way of determining ring membership.  The SMARTS primitive [x] was added by OpenEye and others including RDKit to solve the problem.  [x3] means that the atom has 3 ring bonds incident on it.  That describes an atom at a single ring fusion point.  [x4] means 4 ring bonds, and so includes propellanes.  We probably don't need to go higher than that.
```
m = Chem.MolFromSmiles("C1CC2CCC21")
pat = Chem.MolFromSmarts("[x3,x4]@[x3,x4]")
m.GetSubstructMatches(pat)
```   
<img width="232" alt="image" src="https://github.com/rdkit/rdkit/assets/9198870/9938feb5-20e8-4b9a-8122-cb8c74697ac7">

```
mp = Chem.MolFromSmiles("C1(C2)(C3)C23C1")
mp.GetSubstructMatches(pat)
mp
```
<img width="194" alt="image" src="https://github.com/rdkit/rdkit/assets/9198870/4497b1cf-8480-4935-a853-7a4fb7972f83">

The SMARTS pattern here defines 2 "fusion atoms" joined by a ring bond.  This therefore excludes the bond between two fused ring systems
<img width="346" alt="image" src="https://github.com/rdkit/rdkit/assets/9198870/0159dd23-bc0a-4844-82d7-56a120a1de83">
The jupyter notebook highlights the non-ring bond between the 2 systems, but this is incorrect, as shown by the atom indices returned by GetSubstructMatches.
To count the number of fused rings, you merely need to count the number of lists returned by GetSubstructMatches() for the SMARTS above.

**Reply by bp-kelley** [reply]
To be fair, OpenEye also redefines R to mean ring, not necessarily in the smallest set which is not the RDKit or Daylight specification.

However, the RDKit uses a symmeterized set of smallest rings which avoids some of the behavior @DavidACosgrove is talking about.  That being said, the non standard "x" annotation is probably the way to go here although I would probably use:  

```
pat = Chem.MolFromSmarts("[x&!x1&!x2]")
```

Note that x was added by daylight, so it's in the original smarts specification.

**Reply by DavidACosgrove** [reply]
Yeah, OpenEye’s R is the same as x. It always seemed to me that that was the worst possible way of handling their aversion to SSSR since it meant that the same SMARTS string was interpreted differently in, for example, Daylight and their toolkits (Daylight’s R1 is roughly the same as R2 in OpenEye’s world). I would be happier if they’d just reject a SMARTS string with an R in it. At least then one would have known there was something different going on. 

I think ‘x1’ will be rare in most molecule sets, so the pattern could just be ‘[x;!x2]’. A case could be made for the parser spitting out ‘x1’ as an error.
